### ISSUE TITLE
Allow expunger to run concurrently with transactions

### ISSUE THREAD

**Issue by mdempsky** [urweb/urweb#209]
The current fix for #206 is basically to use a stop-the-world garbage collector. This is simple and correct, but has a measurable impact on latency. E.g., running an HTTP benchmarking tool (e.g., apache bench) against tests/periodicSend will see ~1 second latency spike on requests about once every ~5 seconds when the expunger stops the world to reap clients/channels.

I think it should be possible to avoid this with a more elaborate locking strategy. Broadly, I'm imagining something like:

* Add a global garbage collector generation counter. Also add a generation counter fields to clients.
* Increment the counter every cycle that the expunger runs (or at least every time it actually has anything to expunge).
* At the start of every transaction, read and save the current generation counter.
* Any time a transaction changes a client from `UNUSED` to `USED` or the expunger changes one from `USED` to `UNUSED`, update the generation counter.
* When a transaction accesses a client, only consider it `USED` if the state is `USED` *and* the client's generation count is less than or equal to the global generation counter seen at the start of the transaction.
  * In particular, it's safe to ignore writes to clients with a greater generation count. This just means the client was GC'd after the transaction started, and this is logically no different than if the GC happened immediately after the transaction finished. (As long as expunging mention of the client and its channels from the SQL database is still handled atomically.)

**Comment by achlipala** [comment]
Yeah, something like all that is probably worth implementing, but I don't think your benchmarking scenario is realistic, calling into question why we should optimize for it.  A new client is only allocated when a page is loaded in a browser tab, which probably doesn't happen often enough to build up to one-second latency spikes except for the most diabolically popular apps... right?

**Comment by mdempsky** [comment]
@achlipala Yeah, that's a fair point. I was thinking about this because I was wondering if there's a risk of starving the expunger.

Looking at it somewhat closer, it seems like the expunger is in fact getting starved, and the 1 second latency spikes are actually because the sqlite backend has hard-coded 1 second retries when the database is busy (e.g., as occasionally happens when heavily loaded because of `ab`):

https://github.com/urweb/urweb/blob/2d343807cf2ff4d1ad2a3c9fc550fe4faef4c54b/src/sqlite.sml#L527-L530

**Comment by achlipala** [comment]
Well, that code could deserve improvement, but who really uses SQLite with high concurrency scaling anyway? :)  E.g., what happens if you run the same test with Postgres?

**Comment by mdempsky** [comment]
Repeating the test with postgres is somewhat funny. It's able to handle way more transactions, but also seems much slower at expunging (which is further compounded by the fact that it ends up with more rows to delete).

With periodicSend.ur as committed, I see uw_periodicsend_active grow in rows about 1 per 10 transactions over the course of the `ab` test. I think this is an artifact of `refresh` taking time proportional to the number of rows, and it some knock on effect accumulates over the test.

I tried changing `refresh` to a no-op, the uw_periodicsend_active grows at 1 row per transaction. I did have one 30-second `ab` run where after about 5 seconds / 5000 transactions, the expunger lucked out and acquired the lock, and proceeded to spend the next 25+ seconds slowly expunging all of the clients while transactions were blocked.

I also saw at least one run with sqlite where the expunger got wedged and just started printing "Database is busy" once per second without making any forward progress.

**Comment by achlipala** [comment]
OK, so there are genuine bottlenecks with high load of transactions that allocate new clients, though we wouldn't expect that kind of workload in the wild.  It does still seem worth putting in some modest mechanism to allow more concurrency.
